algorithms/src/fft/polynomial/dense.rs

```diff
@@ -113,10 +112,16 @@ impl<F: Field> DensePolynomial<F> {
         crate::cfg_reduce!(mapping, || zero, |a, b| a + b)
     }
 
-    /// Outputs a polynomial of degree `d` where each coefficient is sampled uniformly at random
-    /// from the field `F`.
+    /// Outputs a univariate polynomial of degree `d` where each non-leading
+    /// coefficient is sampled uniformly at random from R and the leading
+    /// coefficient is sampled uniformly at random from among the non-zero
+    /// elements of R.
     pub fn rand<R: Rng>(d: usize, rng: &mut R) -> Self {
-        let random_coeffs = (0..(d + 1)).map(|_| F::rand(rng)).collect();
+        let mut random_coeffs = (0..(d + 1)).map(|_| F::rand(rng)).collect_vec();
+        while random_coeffs[d].is_zero() {
+            // In the extremely unlikely event, sample again.
+            random_coeffs = (0..(d + 1)).map(|_| F::rand(rng)).collect();
```

Comment:
couldn't we just sample only element `d` here, if only that one is being checked?

Reply:
Neat 9bffffde6